File: go/border/rpkt/path.go
// Copyright 2016 ETH Zurich
// Copyright 2018 ETH Zurich, Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This file handles the path header (parsing/validating/updating/etc).

package rpkt

import (
	"hash"
	"time"

	"github.com/scionproto/scion/go/border/ifstate"
	"github.com/scionproto/scion/go/border/rcmn"
	"github.com/scionproto/scion/go/lib/assert"
	"github.com/scionproto/scion/go/lib/common"
	"github.com/scionproto/scion/go/lib/scmp"
	"github.com/scionproto/scion/go/lib/spath"
	"github.com/scionproto/scion/go/proto"
)

// validatePath validates the path header.
func (rp *RtrPkt) validatePath(dirFrom rcmn.Dir) error {
	// First check if there is a path
	if rp.infoF == nil || rp.hopF == nil {
		return common.NewBasicError("Path required",
			scmp.NewError(scmp.C_Path, scmp.T_P_PathRequired, nil, nil))
	}
	// There is a path, so ifCurr will be set
	if err := rp.validateLocalIF(rp.ifCurr); err != nil {
		return err
	}
	// Check for shorcuts in packets from core links
	if rp.infoF.Shortcut {
		currentLinkType := rp.Ctx.Conf.Net.IFs[*rp.ifCurr].Type
		if currentLinkType == proto.LinkType_core {
			return common.NewBasicError("Shortcut not allowed on core segment", nil)
		}
	}
	// A verify-only Hop Field cannot be used for routing.
	if rp.hopF.VerifyOnly {
		return common.NewBasicError("Hop field is VERIFY_ONLY",
			scmp.NewError(scmp.C_Path, scmp.T_P_NonRoutingHopF, rp.mkInfoPathOffsets(), nil))
	}
	// Check if Hop Field has expired.
	hopfExpiry := rp.infoF.Timestamp().Add(rp.hopF.ExpTime.ToDuration())
	if time.Now().After(hopfExpiry) {
		return common.NewBasicError(
			"Hop field expired",
			scmp.NewError(scmp.C_Path, scmp.T_P_ExpiredHopF, rp.mkInfoPathOffsets(), nil),
			"expiry", hopfExpiry,
		)
	}
	// Verify the Hop Field MAC.
	hfmac := rp.Ctx.Conf.HFMacPool.Get().(hash.Hash)
	err := rp.hopF.Verify(hfmac, rp.infoF.TsInt, rp.getHopFVer(dirFrom))
	rp.Ctx.Conf.HFMacPool.Put(hfmac)
	if err != nil && common.GetErrorMsg(err) == spath.ErrorHopFBadMac {
		err = scmp.NewError(scmp.C_Path, scmp.T_P_BadMac, rp.mkInfoPathOffsets(), err)
	}
	return err
}

// validateLocalIF makes sure a given interface ID exists in the local AS, and
// that it isn't revoked. Note that revocations are ignored if the packet's
// destination is this router.
func (rp *RtrPkt) validateLocalIF(ifid *common.IFIDType) error {
	if ifid == nil {
		return common.NewBasicError("validateLocalIF: Interface is nil", nil)
	}
	if _, ok := rp.Ctx.Conf.Topo.IFInfoMap[*ifid]; !ok {
		// No such interface.
		return common.NewBasicError(
			"Unknown IF",
			scmp.NewError(scmp.C_Path, scmp.T_P_BadIF, rp.mkInfoPathOffsets(), nil),
			"ifid", *ifid,
		)
	}
	for _, e := range rp.HBHExt {
		if e.Type() == common.ExtnOneHopPathType {
			// Ignore revocations if OneHopExtension is present
			return nil
		}
	}
	state, ok := ifstate.LoadState(*ifid)
	if !ok || state.Active {
		// Interface is not revoked
		return nil
	}
	// Interface is revoked.
	sRevInfo := state.SRevInfo
	if sRevInfo == nil {
		rp.Warn("No SRevInfo for revoked interface", "ifid", *ifid)
		return nil
	}
	revInfo, err := sRevInfo.RevInfo()
	if err != nil {
		rp.Warn("Could not parse RevInfo for interface", "ifid", *ifid, "err", err)
		return nil
	}
	err = revInfo.Active()
	if err != nil {
		if !common.IsTimeoutErr(err) {
			rp.Error("Error checking revocation", "err", err)
			return nil
		}
		// If the BR does not have a revocation for the current epoch, it considers
		// the interface as active until it receives a new revocation.
		newState := ifstate.NewInfo(*ifid, true, nil, nil)
		ifstate.UpdateIfNew(*ifid, state, newState)
		return nil
	}
	sinfo := scmp.NewInfoRevocation(
		uint16(rp.CmnHdr.CurrInfoF), uint16(rp.CmnHdr.CurrHopF), uint16(*ifid),
		rp.DirFrom == rcmn.DirExternal, state.RawSRev)
	return common.NewBasicError(
		errIntfRevoked,
		scmp.NewError(scmp.C_Path, scmp.T_P_RevokedIF, sinfo, nil),
		"ifid", ifid,
	)
}

// mkInfoPathOffsets is a helper function to create an scmp.InfoPathOffsets
// instance from the current packet.
func (rp *RtrPkt) mkInfoPathOffsets() scmp.Info {
	return &scmp.InfoPathOffsets{
		InfoF: uint16(rp.CmnHdr.CurrInfoF), HopF: uint16(rp.CmnHdr.CurrHopF),
		IfID: uint16(*rp.ifCurr), Ingress: rp.DirFrom == rcmn.DirExternal,
	}
}

// InfoF retrieves the current path Info Field if it isn't already known.
func (rp *RtrPkt) InfoF() (*spath.InfoField, error) {
	if rp.infoF == nil {
		for _, f := range rp.hooks.Infof {
			ret, infof, err := f()
			switch {
			case err != nil:
				return nil, err
			case ret == HookContinue:
				continue
			case ret == HookFinish:
				rp.infoF = infof
				return rp.infoF, nil
			}
		}
		// Check the common header path metadata validity, and if so extract
		// the current Info Field.
		hOff := rp.CmnHdr.InfoFOffBytes()
		switch {
		case rp.CmnHdr.CurrHopF == rp.CmnHdr.CurrInfoF:
			// There is no path, so do nothing.
		case hOff < rp.idxs.path: // Error
			return nil, common.NewBasicError(
				"Info field index too small",
				scmp.NewError(scmp.C_CmnHdr, scmp.T_C_BadInfoFOffset, nil, nil),
				"min", rp.idxs.path/common.LineLen, "actual", rp.CmnHdr.CurrInfoF,
			)
		case rp.CmnHdr.CurrInfoF > rp.CmnHdr.HdrLen: // Error
			return nil, common.NewBasicError(
				"Info field index too large",
				scmp.NewError(scmp.C_CmnHdr, scmp.T_C_BadInfoFOffset, nil, nil),
				"max", rp.CmnHdr.HdrLen, "actual", rp.CmnHdr.CurrInfoF,
			)
		case rp.CmnHdr.CurrInfoF < rp.CmnHdr.HdrLen: // Parse
			var err error
			if rp.infoF, err = spath.InfoFFromRaw(rp.Raw[hOff:]); err != nil {
				return nil, err
			}
		}
	}
	return rp.infoF, nil
}

// HopF retrieves the current path Hop Field if it isn't already known.
func (rp *RtrPkt) HopF() (*spath.HopField, error) {
	if rp.hopF == nil {
		for _, f := range rp.hooks.HopF {
			ret, hopf, err := f()
			switch {
			case err != nil:
				return nil, err
			case ret == HookContinue:
				continue
			case ret == HookFinish:
				rp.hopF = hopf
				return rp.hopF, nil
			}
		}

		hOff := rp.CmnHdr.HopFOffBytes()
		iOff := rp.CmnHdr.InfoFOffBytes()
		// Check if the common header path metadata is valid, and if so
		// extracts the current Hop Field.
		switch {
		case rp.CmnHdr.CurrHopF == rp.CmnHdr.CurrInfoF:
			// There is no path, so do nothing.
		case hOff < iOff+spath.InfoFieldLength: // Error
			return nil, common.NewBasicError(
				"Hop field index too small",
				scmp.NewError(scmp.C_CmnHdr, scmp.T_C_BadHopFOffset, nil, nil),
				"min", (iOff+spath.InfoFieldLength)/common.LineLen, "actual", rp.CmnHdr.CurrHopF,
			)
		case rp.CmnHdr.CurrHopF >= rp.CmnHdr.HdrLen: // Error
			return nil, common.NewBasicError(
				"Hop field index too large",
				scmp.NewError(scmp.C_CmnHdr, scmp.T_C_BadHopFOffset, nil, nil),
				"max", (rp.CmnHdr.HdrLenBytes()-spath.HopFieldLength)/common.LineLen,
				"actual", rp.CmnHdr.CurrHopF,
			)
		default: // Parse
			var err error
			if rp.hopF, err = spath.HopFFromRaw(rp.Raw[hOff:]); err != nil {
				return nil, err
			}
		}
	}
	return rp.hopF, nil
}

// getHopFVer retrieves the Hop Field (if any) required for verifying the MAC
// of the current Hop Field.
func (rp *RtrPkt) getHopFVer(dirFrom rcmn.Dir) common.RawBytes {
	ingress := dirFrom == rcmn.DirExternal
	var offset int
	if !rp.hopF.Xover || (rp.infoF.Shortcut && !rp.infoF.Peer) {
		offset = rp.getHopFVerNormalOffset()
	} else {
		switch rp.infoF.Peer {
		case true:
			// Peer shortcut paths have two extra HOFs; 1 for the peering
			// interface, and another from the upstream interface, used for
			// verification only.
			switch {
			case ingress && rp.infoF.ConsDir:
				offset = +1
			case ingress && !rp.infoF.ConsDir:
				offset = +2
			case !ingress && rp.infoF.ConsDir:
				offset = -2
			case !ingress && !rp.infoF.ConsDir:
				offset = -1
			}
		case false:
			// Non-peer shortcut paths have an extra HOF above the last hop,
			// used for verification of the last hop in that segment.
			switch {
			case ingress && rp.infoF.ConsDir:
				offset = -1
			case !ingress && !rp.infoF.ConsDir:
				offset = +1
			}
		}
	}
	return rp.hopFVerFromRaw(offset)
}

// getHopFVerNormalOffset is a helper function for getHopFVer, to handle cases
// where the verification Hop Field (if any) is directly before or after
// (depending on the ConsDir flag) the current Hop Field.
func (rp *RtrPkt) getHopFVerNormalOffset() int {
	iOff := rp.CmnHdr.InfoFOffBytes()
	hOff := rp.CmnHdr.HopFOffBytes()
	// If this is the last hop of an Up path, or the first hop of a Down path, there's no previous
	// HOF to verify against.
	if (!rp.infoF.ConsDir &&
		hOff == (iOff+spath.InfoFieldLength+int(rp.infoF.Hops-1)*spath.HopFieldLength)) ||
		(rp.infoF.ConsDir && hOff == (iOff+spath.InfoFieldLength)) {
		return 0
	}
	// Otherwise use the next/prev HOF based on the consDir flag.
	if !rp.infoF.ConsDir {
		return 1
	}
	return -1
}

// hopFVerFromRaw is a helper function for getHopFVer. It extracts the raw
// bytes of the specified Hop Field, excluding the leading flag byte.
func (rp *RtrPkt) hopFVerFromRaw(offset int) common.RawBytes {
	ans := make(common.RawBytes, common.LineLen-1)
	// If the offset is 0, a zero'd slice is returned.
	if offset != 0 {
		b := rp.Raw[(int(rp.CmnHdr.CurrHopF)+offset)*common.LineLen:]
		copy(ans, b[1:common.LineLen])
	}
	return ans
}

// IncPath increments the packet's path, if any. The bool return value is set
// to true if the segment changes (specifically if the packet's metadata is
// updated to the new segment).
func (rp *RtrPkt) IncPath() (bool, error) {
	if rp.infoF == nil {
		// Path is empty, nothing to increment.
		return false, nil
	}
	if assert.On {
		assert.Mustf(rp.consDirFlag != nil, rp.ErrStr, "rp.consDirFlag must not be nil")
	}
	var err error
	var hopF *spath.HopField
	// Initialize to the current InfoF and offset values.
	infoF := rp.infoF
	iOff := rp.CmnHdr.InfoFOffBytes()
	hOff := rp.CmnHdr.HopFOffBytes()
	hdrLen := rp.CmnHdr.HdrLenBytes()
	vOnly := 0
	origConsDir := *rp.consDirFlag
	for {
		hOff += spath.HopFieldLength
		if hOff-iOff > int(infoF.Hops*spath.HopFieldLength) {
			// Passed end of current segment, switch to next segment, and read
			// the new Info Field.
			iOff = hOff
			if infoF, err = spath.InfoFFromRaw(rp.Raw[iOff:]); err != nil {
				// Still return false as the metadata hasn't been updated to the new segment.
				return false, err
			}
			continue
		}
		// Read new Hop Field
		if hopF, err = spath.HopFFromRaw(rp.Raw[hOff:]); err != nil {
			return false, err
		}
		// Find first non-verify-only Hop Field.
		if !hopF.VerifyOnly {
			break
		}
		vOnly++
	}
	if hOff > hdrLen {
		return false, common.NewBasicError("New HopF offset > header length", nil,
			"max", hdrLen, "actual", hOff)
	}
	// Update common header, and packet's InfoF/HopF fields.
	segChgd := iOff != rp.CmnHdr.InfoFOffBytes()
	rp.CmnHdr.UpdatePathOffsets(rp.Raw, uint8(iOff/common.LineLen), uint8(hOff/common.LineLen))
	rp.infoF = infoF
	rp.hopF = hopF
	rp.IncrementedPath = true
	if segChgd {
		// Extract new ConsDir flag.
		rp.consDirFlag = nil
		if _, err = rp.ConsDirFlag(); err != nil {
			return segChgd, err
		}
	}
	// Extract the next interface ID.
	rp.ifNext = nil
	if _, err = rp.IFNext(); err != nil {
		return segChgd, err
	}
	// Check that there's no VERIFY_ONLY fields in the middle of a segment.
	if vOnly > 0 && !segChgd {
		return segChgd, common.NewBasicError("VERIFY_ONLY in middle of segment",
			scmp.NewError(scmp.C_Path, scmp.T_P_BadHopField, rp.mkInfoPathOffsets(), nil))
	}
	// Check that the segment didn't change from a down-segment to an up-segment.
	if origConsDir && !*rp.consDirFlag {
		return segChgd, common.NewBasicError("Switched from down-segment to up-segment",
			scmp.NewError(scmp.C_Path, scmp.T_P_BadSegment, rp.mkInfoPathOffsets(), nil))
	}
	return segChgd, nil
}

// ConsDirFlag retrieves the current path segment's ConsDir flag if not already known.
// (Consdir is defined as being the direction in which the segment was created.)
func (rp *RtrPkt) ConsDirFlag() (*bool, error) {
	if rp.consDirFlag != nil {
		return rp.consDirFlag, nil
	}
	// Try to get ConsDir flag from extensions
	for _, f := range rp.hooks.ConsDirFlag {
		ret, consDir, err := f()
		switch {
		case err != nil:
			return nil, err
		case ret == HookContinue:
			continue
		case ret == HookFinish:
			rp.consDirFlag = &consDir
			return rp.consDirFlag, nil
		}
	}
	// Try to get ConsDir flag from InfoField
	if rp.infoF == nil {
		return nil, nil
	}
	rp.consDirFlag = &rp.infoF.ConsDir
	return rp.consDirFlag, nil
}

// IFCurr retrieves the current interface ID from the packet headers/extensions,
// if not already known.
func (rp *RtrPkt) IFCurr() (*common.IFIDType, error) {
	if rp.ifCurr != nil {
		return rp.ifCurr, nil
	}
	if rp.consDirFlag != nil {
		// Try to get IFID from registered hooks.
		if ifid, err := rp.hookIF(*rp.consDirFlag, rp.hooks.IFCurr); err != nil {
			return nil, err
		} else if ifid != nil {
			return rp.checkSetCurrIF(ifid)
		}
		// Try to get IFID from HopField
		if rp.hopF != nil {
			var ingress bool
			switch rp.DirFrom {
			case rcmn.DirLocal:
				ingress = !*rp.consDirFlag
			case rcmn.DirExternal:
				ingress = *rp.consDirFlag
			default:
				return nil, common.NewBasicError("DirFrom value unsupported", nil,
					"val", rp.DirFrom)
			}
			if ingress {
				return rp.checkSetCurrIF(&rp.hopF.ConsIngress)
			}
			return rp.checkSetCurrIF(&rp.hopF.ConsEgress)
		}
	}
	return nil, nil
}

// checkSetCurrIF is a helper function that ensures the given interface ID is
// valid before setting the ifCurr field and returning the value.
func (rp *RtrPkt) checkSetCurrIF(ifid *common.IFIDType) (*common.IFIDType, error) {
	if ifid == nil {
		return nil, common.NewBasicError("No interface found", nil)
	}
	if _, ok := rp.Ctx.Conf.Net.IFs[*ifid]; !ok {
		return nil, common.NewBasicError("Unknown interface", nil, "ifid", *ifid)
	}
	rp.ifCurr = ifid
	return rp.ifCurr, nil
}

// IFNext retrieves the next interface ID if not already known. As this may be
// an interface in an external ISD-AS, this is not sanity-checked.
func (rp *RtrPkt) IFNext() (*common.IFIDType, error) {
	if rp.ifNext == nil && rp.consDirFlag != nil {
		var err error
		// Try to get IFID from registered hooks.
		if rp.ifNext, err = rp.hookIF(*rp.consDirFlag, rp.hooks.IFNext); err != nil {
			return nil, err
		} else if rp.ifNext != nil {
			return rp.ifNext, nil
		}
		// Get IFID from HopField
		if *rp.consDirFlag {
			rp.ifNext = &rp.hopF.ConsEgress
		} else {
			rp.ifNext = &rp.hopF.ConsIngress
		}
	}
	return rp.ifNext, nil
}

// hookIF is a helper function used by IFCurr/IFNext to run interface ID
// retrival hooks.
func (rp *RtrPkt) hookIF(consDir bool, hooks []hookIntf) (*common.IFIDType, error) {
	for _, f := range hooks {
		ret, intf, err := f(consDir, rp.DirFrom)
		switch {
		case err != nil:
			return nil, err
		case ret == HookContinue:
			continue
		case ret == HookFinish:
			return &intf, nil
		}
	}
	return nil, nil
}

File: python/lib/packet/path.py
# Copyright 2014 ETH Zurich
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
:mod:`path` --- SCION Path packets
==================================
"""
# SCION
from lib.packet.opaque_field import (
    HopOpaqueField,
    InfoOpaqueField,
    OpaqueField,
    OpaqueFieldList,
)
from lib.packet.packet_base import Serializable
from lib.util import Raw


class SCIONPath(Serializable):
    NAME = "SCIONPath"
    A_IOF = "A_segment_iof"
    A_HOFS = "A_segment_hofs"
    B_IOF = "B_segment_iof"
    B_HOFS = "B_segment_hofs"
    C_IOF = "C_segment_iof"
    C_HOFS = "C_segment_hofs"
    OF_ORDER = A_IOF, A_HOFS, B_IOF, B_HOFS, C_IOF, C_HOFS
    IOF_LABELS = A_IOF, B_IOF, C_IOF
    HOF_LABELS = A_HOFS, B_HOFS, C_HOFS

    def __init__(self, raw=None):  # pragma: no cover
        self._ofs = OpaqueFieldList(self.OF_ORDER)
        self._iof_idx = None
        self._hof_idx = None
        super().__init__(raw)

    def _parse(self, raw):
        data = Raw(raw, self.NAME)
        if data:
            # Parse first segment
            a_iof = self._parse_iof(data, self.A_IOF)
            self._parse_hofs(data, self.A_HOFS, a_iof.hops)
        if data:
            # Parse second segment
            b_iof = self._parse_iof(data, self.B_IOF)
            self._parse_hofs(data, self.B_HOFS, b_iof.hops)
        if data:
            # Parse third segment
            assert not a_iof.shortcut
            c_iof = self._parse_iof(data, self.C_IOF)
            self._parse_hofs(data, self.C_HOFS, c_iof.hops)
        self._init_of_idxs()

    def _parse_iof(self, data, label):
        """
        Parse a raw :any:`InfoOpaqueField`.

        :param Raw data: Raw instance.
        :param str label: OF label.
        """
        iof = InfoOpaqueField(data.pop(InfoOpaqueField.LEN))
        self._ofs.set(label, [iof])
        return iof

    def _parse_hofs(self, data, label, count):
        """
        Parse raw :any:`HopOpaqueFields`\s.

        :param Raw data: Raw instance.
        :param str label: OF label.
        :param int count: Number of HOFs to parse.
        """
        hofs = []
        for _ in range(count):
            hofs.append(HopOpaqueField(data.pop(HopOpaqueField.LEN)))
        self._ofs.set(label, hofs)

    @classmethod
    def from_values(cls, a_iof=None, a_hofs=None, b_iof=None,
                    b_hofs=None, c_iof=None, c_hofs=None):  # pragma: no cover
        inst = cls()
        inst._set_ofs(inst.A_IOF, a_iof)
        inst._set_ofs(inst.A_HOFS, a_hofs)
        inst._set_ofs(inst.B_IOF, b_iof)
        inst._set_ofs(inst.B_HOFS, b_hofs)
        inst._set_ofs(inst.C_IOF, c_iof)
        inst._set_ofs(inst.C_HOFS, c_hofs)
        inst._init_of_idxs()
        return inst

    def pack(self):  # pragma: no cover
        raw = self._ofs.pack()
        assert len(raw) == len(self)
        return raw

    def _set_ofs(self, label, value):
        """
        Set an OF label to the given value.

        :param str label: The OF label.
        :param value:
            Can be ``None``, a single Opaque Field, or a list of Opaque Fields.
        """
        if value is None:
            data = []
        elif isinstance(value, list):
            data = value
        else:
            data = [value]
        self._ofs.set(label, data)

    def _init_of_idxs(self):
        self._iof_idx = 0
        self._hof_idx = 0
        if not len(self._ofs):
            return
        iof = self.get_iof()
        if iof.peer:
            hof = self._ofs.get_by_idx(1)
            if hof.xover:
                self._hof_idx += 1
        self.inc_hof_idx()

    def get_of_idxs(self):  # pragma: no cover
        """
        Get current InfoOpaqueField and HopOpaqueField indexes.

        :return: Tuple (int, int) of IOF index and HOF index, respectively.
        """
        return self._iof_idx, self._hof_idx

    def set_of_idxs(self, iof_idx, hof_idx):  # pragma: no cover
        """Set current InfoOpaqueField and HopOpaqueField indexes."""
        self._iof_idx = iof_idx
        self._hof_idx = hof_idx

    def reverse(self):
        """Reverse the direction of the path."""
        if not len(self._ofs):
            # Empty path doesn't need reversal.
            return
        iof_label = self._ofs.get_label_by_idx(self._iof_idx)
        swap_iof, swap_hof = None, None
        # Determine which IOF/HOFs need to be swapped, if any.
        if self._ofs.count(self.C_IOF):
            swap_iof, swap_hof = self.C_IOF, self.C_HOFS
        elif self._ofs.count(self.B_IOF):
            swap_iof, swap_hof = self.B_IOF, self.B_HOFS
        # Do the swap as needed.
        if swap_iof:
            self._ofs.swap(self.A_IOF, swap_iof)
            self._ofs.swap(self.A_HOFS, swap_hof)
        # Reverse IOF flags.
        for label in self.IOF_LABELS:
            self._ofs.reverse_cons_dir_flag(label)
        # Reverse HOF lists.
        for label in self.HOF_LABELS:
            self._ofs.reverse_label(label)
        # Update IOF index:
        # - (1) For paths with a single segment, just get the index of the
        #   original label.
        # - (2) For paths with 2 segments, get the index of the opposite label.
        # - (3) For paths with 3 segments, if the initial label was at either
        #   end, use (2), otherwise use (1), as the current label didn't get
        #   swapped.
        if swap_iof and iof_label == self.A_IOF:
            iof_idx = self._ofs.get_idx_by_label(swap_iof)
        elif swap_iof and iof_label == swap_iof:
            iof_idx = self._ofs.get_idx_by_label(self.A_IOF)
        else:
            iof_idx = self._ofs.get_idx_by_label(iof_label)
        # Update the HOF index by simply subtracting it from the total number of
        # OFs.
        self.set_of_idxs(iof_idx, len(self._ofs) - self._hof_idx)

    def get_hof_ver(self, ingress=True):
        """Return the :any:`HopOpaqueField` needed to verify the current HOF."""
        iof = self.get_iof()
        hof = self.get_hof()
        if not hof.xover or (iof.shortcut and not iof.peer):
            # For normal hops on any type of segment, or cross-over hops on
            # non-peer shortcut hops, just use next/prev HOF.
            return self._get_hof_ver_normal(iof)
        if iof.peer:
            # Peer shortcut paths have two extra HOFs; 1 for the peering
            # interface, and another from the upstream interface, used for
            # verification only.
            ingress_cons_dir = {(True, False): +2, (True, True): +1,
                                (False, False): -1, (False, True): -2}
        else:
            # Non-peer shortcut paths have an extra HOF above the last hop, used
            # for verification of the last hop in that segment.
            ingress_cons_dir = {(True, False): None, (True, True): -1,
                                (False, False): +1, (False, True): None}
        # Map the local direction of travel and the IOF consDir flag to the required
        # offset of the verification HOF (or None, if there's no relevant HOF).
        offset = ingress_cons_dir[ingress, iof.cons_dir_flag]
        if offset is None:
            return None
        return self._ofs.get_by_idx(self._hof_idx + offset)

    def _get_hof_ver_normal(self, iof):
        # If this is the last hop of an Up path, or the first hop of a Down
        # path, there's no previous HOF to verify against.
        if (not iof.cons_dir_flag and self._hof_idx == self._iof_idx + iof.hops) or (
                iof.cons_dir_flag and self._hof_idx == self._iof_idx + 1):
            return None
        # Otherwise use the next/prev HOF based on the consDir flag.
        offset = -1 if iof.cons_dir_flag else 1
        return self._ofs.get_by_idx(self._hof_idx + offset)

    def get_iof(self):  # pragma: no cover
        """Get current :any:`InfoOpaqueField`."""
        if self._iof_idx is None:
            return None
        return self._ofs.get_by_idx(self._iof_idx)

    def get_hof(self):  # pragma: no cover
        """Get current :any:`HopOpaqueField`."""
        if self._hof_idx is None:
            return None
        return self._ofs.get_by_idx(self._hof_idx)

    def inc_hof_idx(self):
        """
        Increment the HOF idx to next routing HOF.

        Skip VERIFY_ONLY HOFs, as they are not used for routing.
        Also detect when there are no HOFs left in the current segment, and
        switch to the next segment, before restarting.
        """
        iof = self.get_iof()
        skipped_verify_only = False
        while True:
            self._hof_idx += 1
            if (self._hof_idx - self._iof_idx) > iof.hops:
                # Switch to the next segment
                self._iof_idx = self._hof_idx
                iof = self.get_iof()
                # Continue looking for a routing HOF
                continue
            hof = self.get_hof()
            if not hof.verify_only:
                break
            skipped_verify_only = True
        return skipped_verify_only

    def get_fwd_if(self):  # pragma: no cover
        """Return the interface to forward the current packet to."""
        if not len(self._ofs):
            return 0
        iof = self.get_iof()
        hof = self.get_hof()
        if iof.cons_dir_flag:
            return hof.egress_if
        return hof.ingress_if

    def get_curr_if(self, ingress=True):  # pragma: no cover
        """
        Return the current interface, depending on the direction of the
        segment.
        """
        hof = self.get_hof()
        iof = self.get_iof()
        if ingress == iof.cons_dir_flag:
            return hof.ingress_if
        return hof.egress_if

    def get_as_hops(self):
        total = 0
        segs = 0
        peer = False
        for l in self.IOF_LABELS:
            res = self._ofs.get_by_label(l)
            if not res:
                break
            peer |= res[0].peer
            total += self._get_as_hops(res[0])
            segs += 1
        if not peer:
            total -= segs - 1
        return total

    def _get_as_hops(self, iof):  # pragma: no cover
        if not iof.shortcut:
            return iof.hops
        if not iof.peer:
            return iof.hops - 1
        return iof.hops - 2

    def is_on_last_segment(self):  # pragma: no cover
        label = self._ofs.get_label_by_idx(self._hof_idx)
        if label == self.A_HOFS:
            return self._ofs.count(self.B_HOFS) == 0
        elif label == self.B_HOFS:
            return self._ofs.count(self.C_HOFS) == 0
        else:
            return True

    def __len__(self):  # pragma: no cover
        """Return the path length in bytes."""
        return len(self._ofs) * OpaqueField.LEN

    def __str__(self):
        s = []
        s.append("<SCION-Path(%sB)>" % len(self))

        for name, iof_label, hofs_label in (
            ("A", self.A_IOF, self.A_HOFS), ("B", self.B_IOF, self.B_HOFS),
            ("C", self.C_IOF, self.C_HOFS),
        ):
            iof = self._ofs.get_by_label(iof_label)
            if not iof:
                break
            s.append("  <%s-Segment>" % name)
            s.append("    %s" % iof[0])
            for of in self._ofs.get_by_label(hofs_label):
                s.append("    %s" % of)
            s.append("  </%s-Segment>" % name)
        s.append("</SCION-Path>")
        return "\n".join(s)


def parse_path(raw):  # pragma: no cover
    return SCIONPath(raw)
